File: sdk/python-sdk/verity_sdk/handlers.py
import logging
from typing import Callable, NewType, Awaitable, Dict

from verity_sdk.utils import Context, unpack_message, MsgType

HandlerFunction = NewType('HandlerFunction', Callable[[str, dict], Awaitable[None]])
DefaultHandlerFunction = NewType('DefaultHandlerFunction', Callable[[dict], Awaitable[None]])


class Handler:
    """
    Defines how to handle a message of a certain type and optionally with a particular status
    """
    def __init__(
            self,
            msg_family: str,
            msg_family_version: str,
            handler_function: HandlerFunction):
        self.msg_family: str = msg_family
        self.msg_family_version: str = msg_family_version
        self.handler_function: HandlerFunction = handler_function

    def handles(self, message: dict) -> bool:

        if '@type' not in message:
            logging.error('message does not contain an "@type" attribute')
            return False
        msg_type = MsgType(message['@type'])
        return msg_type.msg_family == self.msg_family and msg_type.msg_family_version == self.msg_family_version

    async def handle(self, msg_name: str, message: dict):
        """
        Handles the given JSON object message

        Args:
         message (dict): the message to be handled
        """
        await self.handler_function(msg_name, message)

    async def __call__(self, msg_name: str, message: dict):  # Added for clarity and flexibility. Same as `self.handle`.
        await self.handler_function(msg_name, message)


class Handlers:
    """
    Stores an array of message handlers that are used when receiving an inbound message
    """
    _handlers: Dict[str, Handler]
    _default_handler: DefaultHandlerFunction

    def __init__(self):
        self._handlers = {}

    def add_handler(
            self,
            msg_family: str,
            msg_family_version: str,
            handler_function: HandlerFunction):
        """
        Adds a MessageHandler for a message type to the list if current message handlers

        Args:
            msg_family (str): the family of the message to be handled
        msg_family_version (str): the family version to be handled
        handler_function (HandlerFunction):  the handler function itself
        """
        key = Handlers._build_handlers_key(msg_family, msg_family_version)
        self._handlers[key] = Handler(msg_family, msg_family_version, handler_function)

    def has_handler(self, msg_family: str, msg_family_version: str):
        """
        checks list of handlers to see if a handler is register for family

        msg_family (str): the family of the message to be checked
        msg_family_version (str): the family version to be checked

        Returns:
            bool: if there is a handler for given parameters
        """
        return Handlers._build_handlers_key(msg_family, msg_family_version) in self._handlers

    def set_default_handler(self, handler_function: DefaultHandlerFunction):
        """
        Sets a handler to be called for messages where there is no registered handler
        Args:
            handler_function (DefaultHandlerFunction): the function that will be called
        """
        self._default_handler = handler_function

    async def handle_message(self, context: Context, raw_message: bytes):
        """
        Calls all of the handlers that support handling of this particular message type and message status

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
            raw_message (bytes): the raw bytes received from Verity
        """
        message: dict = await unpack_message(context, raw_message)
        msg_type = MsgType(message['@type'])

        if self.has_handler(msg_type.msg_family, msg_type.msg_family_version):
            key = Handlers._build_handlers_key(msg_type.msg_family, msg_type.msg_family_version)
            await self._handlers[key].handle(msg_type.msg_name, message)
        else:
            if self._default_handler is not None:
                await self._default_handler(message)
            else:
                logging.warning('Unable to handle message, and no default handler was defined')

    @staticmethod
    def _build_handlers_key(msg_family: str, msg_family_version: str):
        return '{}{}'.format(msg_family, msg_family_version)

# # Enables handler registration decorator @AddHandler(handlers, msg_family, msg_family_version)
# class AddHandler:
#
#     def __init__(self, handlers: Handlers, msg_family: str, msg_family_version: str,):
#         self.handlers = handlers
#         self.msg_family = msg_family
#         self.msg_family_version = msg_family_version
#
#     def __call__(self, handler_function: HandlerFunction, *args, **kwargs):
#         self.handlers.add_handler(self.msg_family, self.msg_family_version, handler_function)

File: sdk/nodejs-sdk/src/utils/Handlers.js
'use strict'
const utils = require('./index')
const MessageFamily = require('./MessageFamily')

/**
 * Defines how to handle a message of a certain type and optionally with a particular status
 */
class Handler {
  /**
   * Handler
   * @param {String} msgFamily - the family name for the message family
   * @param {String} msgFamilyVersion - the version for the message family
   * @param handlerFunction - function to be called to actually process the message
   */
  constructor (msgFamily, msgFamilyVersion, handlerFunction) {
    this.msgFamily = msgFamily
    this.msgFamilyVersion = msgFamilyVersion
    this.handlerFunction = handlerFunction
  }

  /**
   * Checks to see if this MessageHandler handles a particular agent message
   * @param message the JSON structure of the agent message
   * @return whether or not this Handler handles the given message
   */
  handles (message) {
    if (!('@type' in message)) {
      throw new Error('message does not contain a "@type" attribute')
    }
    const msgType = MessageFamily.parseMessageType(message['@type'])
    return msgType.msgFamily === this.msgFamily && msgType.msgFamilyVersion === this.msgFamilyVersion
  }
}

/**
 * Stores an array of message handlers that are used when receiving an inbound message
 */
class Handlers {
  /** Handlers */
  constructor () {
    this.handlers = {}
    this.defaultHandler = null // function that takes a raw message
  }

  /**
   * Adds a MessageHandler for a message type to the list if current message handlers
   * @param messageFamily the family of the message to be handled
   * @param messageFamilyVersion the message family version
   * @param handlerFunction the handler function itself
   */
  addHandler (msgFamily, msgFamilyVersion, handlerFunction) {
    const handlersKey = Handlers.buildHandlersKey(msgFamily, msgFamilyVersion)
    this.handlers[handlersKey] = new Handler(msgFamily, msgFamilyVersion, handlerFunction)
  }

  /** Checks to see if there is a handler set already
   * @param messageFamily the family of the message to be handled
   * @param messageFamilyVersion the message family version
   * @return bool
   */
  hasHandler (msgFamily, msgFamilyVersion) {
    return Handlers.buildHandlersKey(msgFamily, msgFamilyVersion) in this.handlers
  }

  /**
   * Adds a handler for all message types not handled by other message handlers
   * @param defaultHandlerFunction the function that will be called
   */
  setDefaultHandler (defaultHandlerFunction) {
    this.defaultHandler = defaultHandlerFunction
  }

  /**
   * Calls all of the handlers that support handling of this particular message type and message status
   * @param context an instance of the Context object initialized to a verity-application agent
   * @param rawMessage the raw bytes received from Verity
   */
  async handleMessage (context, rawMessage) {
    const message = await utils.unpackMessage(context, rawMessage)
    const msgType = MessageFamily.parseMessageType(message['@type'])
    const handlersKey = Handlers.buildHandlersKey(msgType.msgFamily, msgType.msgFamilyVersion)

    if (this.hasHandler(msgType.msgFamily, msgType.msgFamilyVersion)) {
      await this.handlers[handlersKey].handlerFunction(msgType.msgName, message)
    } else {
      if (this.defaultHandler) {
        await this.defaultHandler(message)
      }
    }
  }

  /**
   * @param msgFamily the family of the message to be handled
   * @param msgFamilyVersion the message family version
   * @return msgFamily + msgFamilyVersion
   */
  static buildHandlersKey (msgFamily, msgFamilyVersion) {
    return msgFamily + msgFamilyVersion
  }
}

module.exports = {
  Handler: Handler,
  Handlers: Handlers
}
